### PR TITLE
ARROW-5844: [Java] Support comparison & sort for more numeric types

### PR DESCRIPTION
**Pull request by liyafan82**
Currently, we only support comparison & sort for 32-bit integers, in this issue, we provide support for more numeric data types:

byte
short
long
float
double

### CODE REVIEW

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 1]
```diff
@@ -44,6 +85,84 @@ public int compareNotNull(int index1, int index2) {
...
+    public Float4Comparator() {
+      super(Float.SIZE / 8);
+    }
+
+    @Override
+    public int compareNotNull(int index1, int index2) {
+      float value1 = vector1.get(index1);
+      float value2 = vector2.get(index2);
+
+      float result = value1 - value2;
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: I don't think this works for NANs?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 2]
```diff
@@ -44,6 +85,84 @@ public int compareNotNull(int index1, int index2) {
...
+    public Float8Comparator() {
+      super(Double.SIZE / 8);
+    }
+
+    @Override
+    public int compareNotNull(int index1, int index2) {
+      double value1 = vector1.get(index1);
+      double value2 = vector2.get(index2);
+
+      double result = value1 - value2;
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: same comment as above, NaNs?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 3]
```diff
@@ -44,6 +85,84 @@ public int compareNotNull(int index1, int index2) {
...
+    public LongComparator() {
+      super(Long.SIZE / 8);
+    }
+
+    @Override
+    public int compareNotNull(int index1, int index2) {
+      long value1 = vector1.get(index1);
+      long value2 = vector2.get(index2);
+
+      long result = value1 - value2;
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: just return result?

Comment (emkornfield) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: I guess this protects against overflow?

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 1 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Thanks for your kind reminder. In the new commits, NaNs are taken care of.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 2 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Sure. Thanks for the reminder.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 3 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: You are right. This is to protect against overflow which may happen when converting a long to an int.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 4]
```diff
@@ -44,6 +85,110 @@ public int compareNotNull(int index1, int index2) {
...
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]:
```suggestion
      return Long.signum(value1 - value2);
```

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 5]
```diff
@@ -44,6 +85,110 @@ public int compareNotNull(int index1, int index2) {
...
+          return 0;
+        } else if (isNan1) {
+          // nan is greater than any normal value
+          return 1;
+        } else {
+          return -1;
+        }
+      }
+
+      float result = value1 - value2;
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Math.signum()

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 6]
```diff
@@ -44,6 +85,110 @@ public int compareNotNull(int index1, int index2) {
...
+          return 0;
+        } else if (isNan1) {
+          // nan is greater than any normal value
+          return 1;
+        } else {
+          return -1;
+        }
+      }
+
+      double result = value1 - value2;
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: math.signum

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 4 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Good point. Revised accordingly.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 6 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Revised accordingly. Thanks.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java (hunk 5 above)
Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Revised accordingly. Thanks.

File: java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java [hunk 7]
```diff
@@ -44,6 +85,103 @@ public int compareNotNull(int index1, int index2) {
...
+          return 0;
+        } else if (isNan1) {
+          // nan is greater than any normal value
+          return 1;
+        } else {
+          return -1;
+        }
+      }
+
+      float result = Math.signum(value1 - value2);
```
Comment (emkornfield) [inline java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: it seems like you should either directly return the casted signum or use the if statement, not both?

Comment (liyafan82) [reply java/algorithm/src/main/java/org/apache/arrow/algorithm/sort/DefaultVectorComparators.java]: Sure. Thanks a lot.

**Comment by emkornfield** [comment]
+1, LGTM.

**Pull request closed without merging**